### PR TITLE
Send emails to subscribers on new comment

### LINKED ISSUE #467: Mail: modify comment API to send updates when there are subscribers
When a comment is made, we will need to send notifications to any users who are following it. 

This should be behind a feature flag so it isn't released until the rest of the "follow" epic is complete.

### LINKED ISSUE #470: Mail: template for new comment notifications
The email should include the body of the comment and a link to the comments *parent*. If the comment is a top-level comment, then we can simply link to the comment itself. 

Since users may be tempted to reply by e-mail, we will set the reply address to no-reply@discussions.odl.mit.edu (the current reply address)

### PR DESCRIPTION
**Pull request by rhysyngsun**
#### What are the relevant tickets?
Fixes #467 
Fixes #470

#### What's this PR do?
Adds comment notification emails when a comment or post you are subscribed to is commented on.

#### How should this be manually tested?
- Turn on the feature flag `FEATURE_COMMENT_NOTIFICATIONS`
- For two users A and B, ensure both have their comment settings (at `/api/v0/notification_settings/comments/`) set to `"immediate"`
- As user A create a post (you get a subscription for this).
- Comment on that post  as user B and verify user A gets the appropriate email
- Comment on that comment as user A and verify user B gets the appropriate notification email

#### Any background context you want to provide?

There was necessary refactoring of reddit/betamax fixtures up to the top level of the app. These are now conditionally turned on via the `pytest.mark.betamax` marker to provide backwards compatibility with existing cassettes as well as _not_ cause other tests to generate cassettes.

#### Screenshots (if appropriate)
See mockup [here](https://github.com/mitodl/open-discussions/issues/470#issuecomment-379017588)

### CODE REVIEW

**Comment by pdpinch** [comment]
To follow up on our discussion with @Ferdi, please notify the subscriber (to a post) of all comments, no matter the depth.

File: notifications/notifiers/base.py [hunk 1]
```diff
@@ -64,58 +63,56 @@ def can_notify(self, notification_settings, last_notification):
...
+        elif self.notification_settings.is_triggered_weekly:
             trigger_offset = DELTA_ONE_WEEK
         else:
             # practically, we'd only see this if our code called a notifier invalidly for a NEVER trigger_frequency
             raise InvalidTriggerFrequencyError(
-                "Unsupported trigger_frequency: {}".format(notification_settings.trigger_frequency)
+                "Unsupported trigger_frequency: {}".format(self.notification_settings.trigger_frequency)
             )
 
+        print((normalized_now, normalized_created_on, trigger_offset))
```
Comment (noisecapella) [inline notifications/notifiers/base.py]: Stray print statement

File: notifications/notifiers/email.py [hunk 2]
```diff
@@ -57,18 +62,17 @@ def send_notification(self, email_notification):
...
         messages = None
         user = email_notification.user
-        notification_settings = NotificationSettings.objects.get(
-            user=user,
-            notification_type=email_notification.notification_type,
-        )
 
         with utils.mark_as_sent_or_canceled(email_notification) as will_send:
+            if user != self.user:
+                raise Exception("Notification user doesn't match settings user")
```
Comment (noisecapella) [inline notifications/notifiers/email.py]: Are there any cases where this exception would get raised or is this just a precaution?

File: open_discussions/urls.py [hunk 3]
```diff
@@ -33,6 +33,11 @@
     url(r'^$', index, name='open_discussions-index'),
     url(r'^auth_required/$', index),
     url(r'^content_policy/$', index),
+    url(  # so that we can use reverse() to link to this
```
Comment (noisecapella) [inline open_discussions/urls.py]: What were we doing before? Was it entirely handled on the frontend?

File: mail/templates/comments/body.html [hunk 4]
```diff
@@ -0,0 +1,50 @@
+{% extends "email_base.html" %}
+{% load humanize %}
+{% load timeago %}
+
+{% block title %}Open Discussions - New Comment{% endblock %}
+{% block styles %}
+<style type="test/css">
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply mail/templates/comments/body.html]: Can you also make this change on `frontpage/body.html`?

File: mail/templates/frontpage/body.html [hunk 5]
```diff
@@ -32,33 +32,29 @@
...
 </style>
 {% endblock %}
 
 {% block content %}
 
 <div class="section">
-	<p class="lead-in">Here are some recent posts you might have missed</p>
+  <p class="lead-in">Here are some recent posts you might have missed</p>
 </div>
 <table class="recent-posts-table" width="100%" cellpadding="13" cellspacing="0" border="0">
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply mail/templates/frontpage/body.html]: These came from Mailgun's example template and weren't on CSS classes, so I think it's necessary to leave them as attributes.

File: notifications/notifiers/email.py (hunk 2 above)
Comment (rhysyngsun) [reply notifications/notifiers/email.py]: This is a precaution against programmer error.

File: open_discussions/urls.py (hunk 3 above)
Comment (rhysyngsun) [reply open_discussions/urls.py]: Yeah, we basically wildcard matched all urls to the index route and handling it within the react app. Needed to add it here so I could use it in templates.

File: mail/templates/comments/body.html [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+</style>
+{% endblock %}
+
+{% block content %}
+
+<table width="100%" cellpadding="13" cellspacing="0" border="0">
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply mail/templates/comments/body.html]: See my comment below (tldr: I don't think so).

File: notifications/notifiers/base.py [hunk 7]
```diff
@@ -23,38 +18,42 @@
 class BaseNotifier:
     "Base class for notifier"
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply notifications/notifiers/base.py]: Hmm, I wonder if this is something catchable by a linter

File: mail/templates/comments/subject.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+[{{ post.channel_title }}] {% if is_comment_reply %}Reply to your comment "{{ parent.text|truncatechars:20 }}"{% else %} Re: {{ post.title }}{% endif %}
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply mail/templates/comments/subject.txt]: It was in the mockups...So I arbitrarily picked 20.

File: channels/views/comments_test.py [hunk 9]
```diff
@@ -428,6 +425,7 @@ def test_create_comment(client, logged_in_profile):
         'comment_type': 'comment',
         'num_reports': 0,
     }
+    mocked_tasks.notify_subscribed_users.called_once_with(post_id, None, '7')
```
[the remark this answers is not in the document]
Comment (rhysyngsun) [reply channels/views/comments_test.py]: Also, all these assertions should have been on `.delay`

**Comment by rhysyngsun** [comment]
@noisecapella all set

File: notifications/notifiers/email.py [hunk 10]
```diff
@@ -57,18 +62,17 @@ def send_notification(self, email_notification):
...
                 return
 
-            last_notification = self._get_most_recent_notification(notification_settings, before=email_notification)
+            last_notification = self._get_most_recent_notification(before=email_notification)
 
-            data = self._get_notification_data(notification_settings, last_notification)
+            data = self._get_notification_data(email_notification, last_notification)
 
             # generate the message (there's only 1)
             messages = list(api.messages_for_recipients([
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply notifications/notifiers/email.py]: I think this was overlooked

File: notifications/notifiers/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (noisecapella) [reply notifications/notifiers/base.py]: Yes, looks like there's pydocstyle: https://pypi.org/project/pydocstyle/

[the remark this answers is not in the document]
Comment (noisecapella) [reply notifications/notifiers/base.py]: Made an issue for this: #623

**Comment by noisecapella** [comment]
Actually there is this minor change: https://github.com/mitodl/open-discussions/pull/561#discussion_r182450602

**Comment by rhysyngsun** [comment]
@noisecapella fixed that and another docstring inconsistency I spotted in the same file

**Comment by noisecapella** [comment]
:+1:

**Pull request merged**